quickwit-metastore/src/metastore/single_file_metastore.rs

```diff
@@ -0,0 +1,1352 @@
+
+            // Check for the existence of index.
+            if data.contains_key(&index_uri) {
+                return Err(MetastoreErrorKind::ExistingIndexUri
+                    .with_error(anyhow::anyhow!("Index already exists: {:?}", &index_uri)));
+            }
+
+            // Insert metadata set.
+            let metadata_set = MetadataSet {
+                index: IndexMetadata {
+                    version: FILE_FORMAT_VERSION.to_string(),
+                },
+                splits: HashMap::new(),
+            };
+            data.insert(index_uri.clone(), metadata_set.clone());
+
+            // Serialize metadata set.
+            contents = match serde_json::to_vec(&metadata_set) {
+                Ok(c) => c,
+                Err(e) => {
+                    return Err(MetastoreErrorKind::InvalidManifest
+                        .with_error(anyhow::anyhow!("Failed to serialize meta data: {:?}", e)));
+                }
+            };
+        }
+
+        // Put data back into storage.
+        self.storage
+            .put(&Path::new(&index_uri), PutPayload::from(contents))
+            .await
+            .map_err(|e| {
+                MetastoreErrorKind::InternalError
+                    .with_error(anyhow::anyhow!("Failed to put metadata set: {:?}", e))
+            })?;
+
+        Ok(())
+    }
+
+    async fn delete_index(&self, index_uri: IndexUri) -> MetastoreResult<()> {
+        {
+            let mut data = self.data.write().unwrap();
+
+            // Check for the existence of index.
+            if !data.contains_key(&index_uri) {
+                return Err(MetastoreErrorKind::IndexDoesNotExist
+                    .with_error(anyhow::anyhow!("Index does not exist: {:?}", &index_uri)));
+            }
+
+            // Remove metadata set.
+            data.remove(&index_uri);
+        }
+
+        // Delete data from storage.
+        self.storage
+            .delete(&Path::new(&index_uri))
+            .await
+            .map_err(|e| {
+                MetastoreErrorKind::InternalError
+                    .with_error(anyhow::anyhow!("Failed to delete metadata set: {:?}", e))
+            })?;
+
+        Ok(())
+    }
+
+    async fn stage_split(
+        &self,
+        index_uri: IndexUri,
+        split_id: SplitId,
+        mut split_metadata: SplitMetadata,
+    ) -> MetastoreResult<SplitId> {
+        // Overwrite split state to Staged.
+        split_metadata.split_state = SplitState::Staged;
+
+        // Serialized metadata set.
+        let contents: Vec<u8>;
+
+        {
+            let mut data = self.data.write().unwrap();
+
+            // Check for the existence of index.
+            if !data.contains_key(&index_uri) {
+                return Err(MetastoreErrorKind::IndexDoesNotExist
+                    .with_error(anyhow::anyhow!("Index does not exist: {:?}", &index_uri)));
+            }
+
+            let metadata_set = data.get_mut(&index_uri).unwrap();
```

Comment:
Whenever possible avoid unwrap...

It makes it much easier to make it clear that panicking is not possible.

Here is an idiom that allows you to avoid the Option unwrap.
```rust
let metadata_set = data.get_mut(&index_uri)
    .ok_or_else(|| ...)?;
```

Note your code will actually end up shorter.

If you are not fond of those functional API, it is possible also to do...

```
let metadata_set = 
      if let Some(metadata_set) = ... {
         metadata_set     
     } else {
         return Err(...);
      }
```

Reply:
Also fixed a similar code.